// vite-plugin-dynamic-import/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fixGlob = exports.cleanUrl = exports.hasDynamicImport = exports.hashRE = exports.queryRE = exports.dynamicImportRegex = exports.multilineCommentsRegex = exports.singleCommentsRegex = exports.KNOWN_SFC_EXTENSIONS = exports.JS_EXTENSIONS = void 0;
exports.JS_EXTENSIONS = [
    '.mjs',
    '.js',
    '.ts',
    '.jsx',
    '.tsx',
    '.cjs'
];
exports.KNOWN_SFC_EXTENSIONS = [
    '.vue',
    '.svelte',
];
exports.singleCommentsRegex = /([^\:])\/\/.*/g;
exports.multilineCommentsRegex = /\/\*(.|[\r\n])*?\*\//gm;
exports.dynamicImportRegex = /\bimport[\s\r\n]*?\(/;
exports.queryRE = /\?.*$/s;
exports.hashRE = /#.*$/s;
function hasDynamicImport(code) {
    code = code
        .replace(exports.singleCommentsRegex, '')
        .replace(exports.multilineCommentsRegex, '');
    return exports.dynamicImportRegex.test(code);
}
exports.hasDynamicImport = hasDynamicImport;
const cleanUrl = (url) => url.replace(exports.hashRE, '').replace(exports.queryRE, '');
exports.cleanUrl = cleanUrl;
/**
 * ```
 * 🚧 in some cases, glob may not be available
 * e.g. fill necessary slash `../views*` -> `../views/*`
 * ```
 */
function fixGlob(glob) {
    const [, importPath] = glob.match(/(.*\w\/?)\*/);
    if (!importPath.endsWith('/')) {
        return glob.replace(importPath, importPath + '/');
    }
}
exports.fixGlob = fixGlob;
